htlcswitch/link.go

```diff
@@ -2503,12 +2613,16 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
 	}, false)
 }
 
-// fail helper function which is used to encapsulate the action necessary for
-// proper disconnect.
-func (l *channelLink) fail(format string, a ...interface{}) {
+// fail is a function which is used to encapsulate the action necessary for
+// properly failing the link. It takes a LinkFailureError, which will be passed
+// to the OnChannelFailure closure, in order for it to determine if we should
+// force close the channel, and if we should send an error message to the
+// remote peer.
+func (l *channelLink) fail(linkErr LinkFailureError,
+	format string, a ...interface{}) {
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
+	l.errorf("Failing link: %s", reason)
+	l.cfg.OnChannelFailure(l.ChanID(), l.ShortChanID(), linkErr)
```

Comment:
If the `Disconnect` method on the peer is no longer used, we may want to consider removing it all together. This isn't a blocker for this PR though, and can be done at a later time.

Reply:
Removed.